api/src/main/java/org/openmrs/api/impl/ConceptServiceImpl.java

```diff
@@ -1982,6 +1984,72 @@ public boolean hasAnyConceptAttribute(ConceptAttributeType conceptAttributeType)
 		return dao.getConceptAttributeCount(conceptAttributeType) > 0;
 	}
 
+	/**
+	 * @see org.openmrs.api.ConceptService#getConceptByNameOrIdOrMap(String conceptRef
+	 */
+	@Override
+	@Transactional(readOnly = true)
+	public Concept getConceptByNameOrIdOrMap(String conceptRef) throws APIException {
+		if (conceptRef == null) {
+			return null;
+		}
+		Concept cpt = null;
+		//check if input is a valid Uuid
+		if (!(conceptRef.length() < 36 || conceptRef.length() > 38 || conceptRef.contains(" ") || conceptRef.contains("."))) {
+			cpt = Context.getConceptService().getConceptByUuid(conceptRef);
+			if (cpt != null)
+				return cpt;
+		}
+		//check if input is map
+		if (conceptRef.indexOf(":") != -1) {
+			int idx = conceptRef.indexOf(":");
+			String conceptSource, conceptCode;
+			conceptCode = conceptRef.substring(idx + 1);
+			conceptSource = conceptRef.substring(0, idx);
+			cpt = Context.getConceptService().getConceptByMapping(conceptCode, conceptSource);
+			if (cpt != null)
+				return cpt;
+		}
+		//check if input is a name or id
+		if (!conceptRef.isEmpty()) {
+			//handle id
+			if (NumberUtils.isParsable(conceptRef)) {
+				int conceptId = Integer.valueOf(conceptRef);
```

Comment:
Again, this parses the string twice for no real value.

```java
int conceptId = NumberUtils.toInt(conceptRef, -1);
if (conceptId >= 0) {
    cpt = Context.getConceptService().getConcept(conceptId);
}
```

Reply:
You are magical!